File: competitive-programming/src/edu/algos/li02_hashing/L217_ContainsDuplicate.java
package edu.algos.li02_hashing;

import java.util.HashSet;

public class L217_ContainsDuplicate {

	public static void main(String[] args) {
		L217_ContainsDuplicate obj = new L217_ContainsDuplicate();
		System.out.println(obj.containsDuplicate(new int[] {1,2,3,1}));
		System.out.println(obj.containsDuplicate(new int[] {1,2,3,4}));
	}
	
    // Problem: https://leetcode.com/problems/contains-duplicate/
    // Idea: Use Set to check duplicates
    public boolean containsDuplicate(int[] nums) {
        HashSet<Integer> set = new HashSet<Integer>();
        for(int i=0; i<nums.length; i++){
            if(set.contains(nums[i])){
                return true;
            }
            set.add(nums[i]);
        }
        return false;
    }
}

File: dsa/patterns/02_hashing/python/l217_contains_duplicate.py
"""
L217 - Contains Duplicate

Problem: https://leetcode.com/problems/contains-duplicate/
Idea: Use Set to check duplicates
Time: O(n)
Space: O(n)
"""

def contains_duplicate(nums):
    seen = set()
    
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    
    return False


if __name__ == "__main__":
    test_cases = [
        [1, 2, 3, 1],
        [1, 2, 3, 4],
        [1, 1, 1, 3, 3, 4, 3, 2, 4, 2]
    ]
    
    for i, nums in enumerate(test_cases, 1):
        print(f"Test case {i}: {nums}")
        print(f"Contains duplicate: {contains_duplicate(nums)}")
        print()
